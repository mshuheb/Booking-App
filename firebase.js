import { getApp, initializeApp } from "firebase/app";
import { initializeAuth, getReactNativePersistence } from "firebase/auth";
import { getFirestore } from "firebase/firestore";
import ReactNativeAsyncStorage from "@react-native-async-storage/async-storage";

const firebaseConfig = {
  apiKey: "AIzaSyBRqeTVnFdloEn3bpB9fKk_btNd0tPKWok",
  authDomain: "booking-project-afa36.firebaseapp.com",
  projectId: "booking-project-afa36",
  storageBucket: "booking-project-afa36.appspot.com",
  messagingSenderId: "1096229966164",
  appId: "1:1096229966164:web:c1238ddd012239251e38c1",
};

let app;
try {
  app = getApp();
} catch {
  app = initializeApp(firebaseConfig);
}

const auth = initializeAuth(app, {
  persistence: getReactNativePersistence(ReactNativeAsyncStorage),
});

const db = getFirestore();

export { auth, db };
