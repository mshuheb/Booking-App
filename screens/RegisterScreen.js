import { Text, View, SafeAreaView, KeyboardAvoidingView, Pressable, TextInput, Alert, ActivityIndicator } from 'react-native';
import React, { useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import { createUserWithEmailAndPassword } from 'firebase/auth';
import { auth, db } from '../firebase';
import { setDoc, doc } from 'firebase/firestore';
import AsyncStorage from '@react-native-async-storage/async-storage';

const RegisterScreen = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [phone, setPhone] = useState('');
  const [isLoading, setIsLoading] = useState(false); 
  const navigation = useNavigation();

  const emailRegex = /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,6}$/; 
  const phoneRegex = /^[0-9]{10}$/; 

  const register = async () => {
    if (email === '' || password === '' || phone === '') {
      Alert.alert(
        'Invalid Details',
        'Please enter all the credentials',
        [{ text: 'OK' }],
        { cancelable: false }
      );
      return;
    }

    if (!emailRegex.test(email)) {
      Alert.alert('Invalid Email', 'Please enter a valid email address.', [
        { text: 'OK' },
      ]);
      return;
    }

    if (!phoneRegex.test(phone)) {
      Alert.alert('Invalid Phone', 'Please enter a valid 10-digit phone number.', [
        { text: 'OK' },
      ]);
      return;
    }

    setIsLoading(true); 

    try {
      const userCredentials = await createUserWithEmailAndPassword(auth, email, password);
      const user = userCredentials.user.email;
      const uid = auth.currentUser.uid;

      
      await setDoc(doc(db, 'users', uid), {
        email: user,
        phone: phone,
      });

      
      await AsyncStorage.setItem('userEmail', user);
      await AsyncStorage.setItem('userPhoneNumber', phone);

      
      Alert.alert('Success', 'Registration successful!', [
        {
          text: 'OK',

          onPress: () => {
            setEmail('');
            setPassword('');
            setPhone('');
            navigation.navigate('Login');
          },
        },
      ]);
    } catch (error) {
      console.error('Registration error:', error);
      Alert.alert('Registration Failed', 'Something went wrong. Please try again.', [
        { text: 'OK' },
      ]);
    } finally {
      setIsLoading(false); 
    }
  };

  return (
    <SafeAreaView style={{ flex: 1, backgroundColor: 'white', padding: 10, alignItems: 'center' }}>
      <KeyboardAvoidingView>
        <View style={{ justifyContent: 'center', alignItems: 'center', marginTop: 100 }}>
          <Text style={{ color: '#003580', fontSize: 17, fontWeight: '700' }}>Register</Text>
          <Text style={{ marginTop: 15, fontSize: 18, fontWeight: '500' }}>Create an Account</Text>
        </View>

        <View style={{ marginTop: 50 }}>
          <View>
            <Text style={{ fontSize: 18, fontWeight: '600', color: 'gray' }}>Email</Text>
            <TextInput
              value={email}
              onChangeText={(text) => setEmail(text)}
              placeholder="Enter your email id"
              placeholderTextColor={'black'}
              style={{
                fontSize: email ? 18 : 18,
                borderBottomColor: 'gray',
                borderBottomWidth: 1,
                marginVertical: 10,
                width: 300,
              }}
            />
          </View>

          <View style={{ marginTop: 15 }}>
            <Text style={{ fontSize: 18, fontWeight: '600', color: 'gray' }}>Password</Text>
            <TextInput
              value={password}
              onChangeText={(text) => setPassword(text)}
              secureTextEntry={true}
              placeholder="Password"
              placeholderTextColor={'black'}
              style={{
                fontSize: password ? 18 : 18,
                borderBottomColor: 'gray',
                borderBottomWidth: 1,
                marginVertical: 10,
                width: 300,
              }}
            />
          </View>

          <View style={{ marginTop: 15 }}>
            <Text style={{ fontSize: 18, fontWeight: '600', color: 'gray' }}>Phone</Text>
            <TextInput
              value={phone}
              onChangeText={(text) => setPhone(text)}
              placeholder="Enter your Phone No"
              placeholderTextColor={'black'}
              style={{
                fontSize: phone ? 18 : 18,
                borderBottomColor: 'gray',
                borderBottomWidth: 1,
                marginVertical: 10,
                width: 300,
              }}
            />
          </View>
        </View>

        {isLoading ? ( 
          <ActivityIndicator size="large" color="#003580" style={{ marginTop: 50 }} />
        ) : (
          <Pressable
            onPress={register}
            style={{
              width: 200,
              backgroundColor: '#003580',
              padding: 15,
              borderRadius: 7,
              marginTop: 50,
              marginLeft: 'auto',
              marginRight: 'auto',
            }}
            testID="registerButton"
          >
            <Text style={{ textAlign: 'center', color: 'white', fontSize: 17, fontWeight: 'bold' }}>Register</Text>
          </Pressable>
        )}

        <Pressable onPress={() => navigation.goBack()} style={{ marginTop: 20 }}>
          <Text style={{ textAlign: 'center', color: 'gray', fontSize: 17 }}>
            Already have an account? Sign In
          </Text>
        </Pressable>
      </KeyboardAvoidingView>
    </SafeAreaView>
  );
};

export default RegisterScreen;
