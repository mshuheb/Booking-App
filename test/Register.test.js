import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { Alert } from "react-native";
import RegisterScreen from "../screens/RegisterScreen";
import { createUserWithEmailAndPassword } from "firebase/auth";
import { useNavigation } from "@react-navigation/native";


jest.mock("firebase/auth", () => ({
  createUserWithEmailAndPassword: jest.fn(),
  initializeAuth: jest.fn().mockReturnValue({}),
  getReactNativePersistence: jest.fn(),
}));

jest.mock("@react-native-async-storage/async-storage", () => ({
  setItem: jest.fn(),
  getItem: jest.fn(),
  removeItem: jest.fn(),
}));

jest.mock("@react-navigation/native", () => ({
  useNavigation: jest.fn(),
}));

jest.spyOn(Alert, "alert").mockImplementation(() => {});

describe("<RegisterScreen />", () => {
  let navigationMock;

  beforeEach(() => {
    navigationMock = { navigate: jest.fn(), goBack: jest.fn() };
    useNavigation.mockReturnValue(navigationMock);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("renders correctly with initial UI elements", () => {
    const { getByTestId, getByText, getByPlaceholderText } = render(<RegisterScreen />);
  
    expect(getByText("Create an Account")).toBeTruthy();
    expect(getByPlaceholderText("Enter your email id")).toBeTruthy();
    expect(getByPlaceholderText("Password")).toBeTruthy();
    expect(getByPlaceholderText("Enter your Phone No")).toBeTruthy();
    
    expect(getByTestId("registerButton")).toBeTruthy();
  });

  test("matches snapshot", () => {
    const { toJSON } = render(<RegisterScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  test("validates empty form fields and shows an alert", async () => {
    const { getByTestId } = render(<RegisterScreen />);

    const registerButton = getByTestId("registerButton");
  
    fireEvent.press(registerButton);
  
    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(
        "Invalid Details",
        "Please enter all the credentials",
        [{ text: "OK" }],
        { cancelable: false }
      );
    });
  });

  test("navigates back to the Login screen on 'Already have an account?' press", () => {
    const { getByText } = render(<RegisterScreen />);
    const signInText = getByText("Already have an account? Sign In");

    fireEvent.press(signInText);

    expect(navigationMock.goBack).toHaveBeenCalled();
  });

  test("shows an alert when an invalid email is entered", async () => {
    const { getByTestId, getByPlaceholderText } = render(<RegisterScreen />);
  
    fireEvent.changeText(getByPlaceholderText("Enter your email id"), "invalidemail");
    fireEvent.changeText(getByPlaceholderText("Password"), "password123");
    fireEvent.changeText(getByPlaceholderText("Enter your Phone No"), "1234567890");
  
    fireEvent.press(getByTestId("registerButton"));
  
    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(
        "Invalid Email",
        "Please enter a valid email address.",
        [{ text: "OK" }]
      );
    });
  });

  test("shows an alert when an invalid phone number is entered", async () => {
    const { getByTestId, getByPlaceholderText } = render(<RegisterScreen />);
  
    fireEvent.changeText(getByPlaceholderText("Enter your email id"), "test@example.com");
    fireEvent.changeText(getByPlaceholderText("Password"), "password123");
    fireEvent.changeText(getByPlaceholderText("Enter your Phone No"), "12345"); 
  
    fireEvent.press(getByTestId("registerButton"));
  
    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(
        "Invalid Phone",
        "Please enter a valid 10-digit phone number.",
        [{ text: "OK" }]
      );
    });
  });

  test("shows an error alert when registration fails", async () => {
    createUserWithEmailAndPassword.mockRejectedValueOnce(new Error("Network error"));
  
    const { getByTestId, getByPlaceholderText } = render(<RegisterScreen />);
  
    fireEvent.changeText(getByPlaceholderText("Enter your email id"), "test@example.com");
    fireEvent.changeText(getByPlaceholderText("Password"), "password123");
    fireEvent.changeText(getByPlaceholderText("Enter your Phone No"), "1234567890");
  
    fireEvent.press(getByTestId("registerButton"));
  
    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(
        "Registration Failed",
        "Something went wrong. Please try again.",
        [{ text: "OK" }]
      );
    });
  });
});
