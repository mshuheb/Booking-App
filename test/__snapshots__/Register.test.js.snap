// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterScreen /> matches snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "padding": 10,
    }
  }
>
  <View
    onLayout={[Function]}
  >
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "marginTop": 100,
        }
      }
    >
      <Text
        style={
          {
            "color": "#003580",
            "fontSize": 17,
            "fontWeight": "700",
          }
        }
      >
        Register
      </Text>
      <Text
        style={
          {
            "fontSize": 18,
            "fontWeight": "500",
            "marginTop": 15,
          }
        }
      >
        Create an Account
      </Text>
    </View>
    <View
      style={
        {
          "marginTop": 50,
        }
      }
    >
      <View>
        <Text
          style={
            {
              "color": "gray",
              "fontSize": 18,
              "fontWeight": "600",
            }
          }
        >
          Email
        </Text>
        <TextInput
          onChangeText={[Function]}
          placeholder="Enter your email id"
          placeholderTextColor="black"
          style={
            {
              "borderBottomColor": "gray",
              "borderBottomWidth": 1,
              "fontSize": 18,
              "marginVertical": 10,
              "width": 300,
            }
          }
          value=""
        />
      </View>
      <View
        style={
          {
            "marginTop": 15,
          }
        }
      >
        <Text
          style={
            {
              "color": "gray",
              "fontSize": 18,
              "fontWeight": "600",
            }
          }
        >
          Password
        </Text>
        <TextInput
          onChangeText={[Function]}
          placeholder="Password"
          placeholderTextColor="black"
          secureTextEntry={true}
          style={
            {
              "borderBottomColor": "gray",
              "borderBottomWidth": 1,
              "fontSize": 18,
              "marginVertical": 10,
              "width": 300,
            }
          }
          value=""
        />
      </View>
      <View
        style={
          {
            "marginTop": 15,
          }
        }
      >
        <Text
          style={
            {
              "color": "gray",
              "fontSize": 18,
              "fontWeight": "600",
            }
          }
        >
          Phone
        </Text>
        <TextInput
          onChangeText={[Function]}
          placeholder="Enter your Phone No"
          placeholderTextColor="black"
          style={
            {
              "borderBottomColor": "gray",
              "borderBottomWidth": 1,
              "fontSize": 18,
              "marginVertical": 10,
              "width": 300,
            }
          }
          value=""
        />
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#003580",
          "borderRadius": 7,
          "marginLeft": "auto",
          "marginRight": "auto",
          "marginTop": 50,
          "padding": 15,
          "width": 200,
        }
      }
      testID="registerButton"
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 17,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Register
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "marginTop": 20,
        }
      }
    >
      <Text
        style={
          {
            "color": "gray",
            "fontSize": 17,
            "textAlign": "center",
          }
        }
      >
        Already have an account? Sign In
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
